### ISSUE TITLE
Move cannot be converted

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-engine#847]
Starting model:

```html
<messageQuote>
  <paragraph>abc</paragraph>
  <paragraph>xx</paragraph>
  <paragraph>def</paragraph>
</messageQuote>
```

Starting view:

```html
<div><blockquote><p>abc</p><p>xx</p><p>def</p></blockquote></div>
```

I want to unwrap that `<messageQuote>`. It first triggers move of its content before it:

```html
<paragraph>abc</paragraph>
<paragraph>xx</paragraph>
<paragraph>def</paragraph>
<messageQuote></messageQuote>
```

From what I can see, the `remove()` model-to-view-converter is called with the `sourcePostion`: `[0, 0]` and `offsetSize`: `1` which should map to:

```html
<div><blockquote>[<p>abc</p>]<p>xx</p><p>def</p></blockquote></div>
```

But apparently converts to:

```html
<div><blockquote><p>{a}bc</p><p>xx</p><p>def</p></blockquote></div>
```

;|

I can see that at that moment, the model looks like this already:

```html
<paragraph>abc</paragraph>
<paragraph>xx</paragraph>
<paragraph>def</paragraph>
<messageQuote></messageQuote>
```

So the `[ 0, 0 ]` position + 1 offset is clearly `<paragraph>[a]bc</paragraph>...`. Apparently, the `sourcePosition` needs to be transformed before move is converted.

**Comment by Reinmar** [comment]
I guess I should prepare the `sourcePosition` in `convertMove()`, but I don't know how :D @scofalik, heeelp!

**Comment by scofalik** [comment]
Hard to say without seeing any code. You are using `batch().unwrap()`?

BTW. When you are converting `remove()`, your view should already have `move()` converted, and be `<div><p>{a}bc</p><p>xx</p><p>def</p><blockquote></blockquote></div>`.

I've checked `UnwrapDelta` and it seems correct. `UnwrapDelta` is a move + remove. From what I can see in your example it seems that the move is correct. What you have to check is:

1. If `RemoveOperation` had correct values and returned correct data after `_execute`.
2. If model state is correct (3 `paragraph`s in main root, `messageQuote` in graveyard).
3. If proper values are passed to `remove()` converter. They are passed in `data` parameter. Basically, you should check `sourcePosition` and `data.item`.

Looking at your example, it seems that something is in fact wrong with `sourcePosition`. `sourcePosition` is taken directly from `RemoveOperation#_execute` returned value.

To sum up, first find where the problem really lies.

**Comment by Reinmar** [comment]
Adding this helped:

```js

		// In the model the insertion already happened, but we need to pass to convertRemove() postion
		// as nothing happened yet.
		const sourcePositionBeforeInsertion
			= sourcePosition._getTransformedByInsertion( range.start, range.end.offset - range.start.offset );
```

WIll make a PR.

**Comment by Reinmar** [comment]
Turns out this was far from complete. The problem is that `sourcePosition`'s parent in the model has changed already due to the insertion which will be converted later, but already took place in the model.

This means that we can't map it to the view, because we need the previous parent of it.

So the idea is to convert changes in the source order in which they happened in the model, not the order in which they were applied. This will ensure that a source position we are currently converting is correct because none of the other model changes has changed its parent yet.

The only dubious moment here is that if we're moving `<p>` before the `<x>`:

```html
<x>a</x><p>b</p> -> <p>b</p><x>a</x>
```

First, we'll convert inserting it. So for a time being, we'll have two `<p>`s in the view. 

This shouldn't be a problem because to map positions between  model and view we map parents of them so the root element in this case. The offset of `<x>` will need to be fixed (when removing it) but the rest will be fine.

**Comment by Reinmar** [comment]
How it'll work in the case I had initially:

```html
<messageQuote><paragraph>a[bc</paragraph><paragraph>de]f</paragraph></messageQuote>
```

We want to unwrap the `<mQ>` (remove it).

So first we convert insertion of 2 paragraphs:

```html
<p>abc</p><p>def</p><blockquote><p>abc</p><p>def</p></blockquote>
```

To do that we have to convert insert position at `[0]` and it correctly maps to the view because the root hasn't changed.

Then we need to convert remove of 2 items at the old position `[0,0]` which we need to transform (by insertion) to new position  `[2,0]` which has a correct `parent == messageQuote` in the model which maps well to `blockquote` in the view.

Ufff!
